### DISCUSSION TITLE
Proposal: Add syntax support to CurrentCulture/InvariantCulture for interpolated strings

### DISCUSSION THREAD

**Discussion by iam3yal** [dotnet/csharplang#177] [General]
_Disclaimer: I've copy/paste this from [#12336](https://github.com/dotnet/roslyn/issues/12336) but did some modifications._

## Summary
[summary]: #summary

Add syntax support to `CurrentCulture`/`InvariantCulture` for interpolated strings.

## Motivation
[motivation]: #motivation

C# introduced interpolated string but with no obvious way to decide which `CultureInfo` to use during formatting.

Currently it is possible to use `InvariantCulture` like this:

```
FormattableString.Invariant($"count: {count}")
```

This requires code to target .NET 4.6 and allocates the `FormatabbleString` instance instead of just generating call to string.Format.

## Detailed design
[design]: #detailed-design

This could be improved by adding prefixes to explicitly choose `CultureInfo` like this:

```
$C"count: {count}"    // This will call string.Format with CultureInfo.CurrentCulture
$@C"count: {count}"   // Same as above for verbatim string literals 
$I"count: {count}"    // This will call string.Format with CultureInfo.InvariantCulture
$@I"count: {count}"   // Same as above for verbatim string literals 
```

This will compile to `string.Format` and will not require targeting .NET 4.6.

Additionally, it will be easy to add analyzer to enforce setting the `C` and `I` prefixes always, like we have now in FxCop for string.Format and others.

p.s. I think that capital letters standout better but maybe lowercase letters might be slightly easier to type.

**Comment by dlemstra** [comment]
I like what is done in ECMAScript 6 and I would like to purpose a slightly different syntax than what is written above. How about the following syntax (with a backtick instead of the $):

```
C`count: {count}`    // This will call string.Format with CultureInfo.CurrentCulture
@C`count: {count}`   // Same as above for verbatim string literals 
I`count: {count}`    // This will call string.Format with CultureInfo.InvariantCulture
@I`count: {count}`   // Same as above for verbatim string literals 

```   

This syntax will also make is easier to use a quote inside the string because you no longer need to replace them:

```
@`<html>
<div data-count="{count}"/>
</html>`
```

**Comment by HaloFour** [comment]
@dlemstra 

String interpolation already uses `$` in C#, I don't think it makes any sense to incorporate a second syntax borrowed from another language.

**Comment by alrz** [comment]
On syntax, the leading `$` is mainly used for lexer to expect an interpolated string. culture invariance could be also an optional suffix e.g. `$@"{e}"i` since it would not affect lexer.

**Comment by jnyrup** [comment]
@dlemstra from _The C# Programming Language_ 
> In the early days of the C# language design, we experimented
with using the backtick (`) character. Personally, I really liked this choice: The backtick
is in the quote family and is a grossly underused character. In fact, it was so underused
that some international keyboards don’t even include a key for it. We were lucky to
have such a good, early international following for C#, which allowed us to fix this
potential error very early in the design of the language.
-- Brad Abrams

I can type a backtick on my keyboard (Danish QWERTY) but it is a [dead key](https://en.wikipedia.org/wiki/Dead_key) which means I have to type `Shift+[backtick key], Space` to get a backtick -- or a grave accented whitespace.

This most likely takes the prize as the common programming character most difficult for me to type.

**Comment by orthoxerox** [comment]
@alrz implementing the flag as a postfix will affect the most important lexer, the human lexer 😀. If your string is long enough you'll have to scan till its end to find out if the treatment is invariant. It can already be annoying with regexps, but regexps are (usually) short enough for this to work.

**Comment by jnm2** [comment]
Yeah, I like the prefix `$i` best.

**Comment by dstarkowski** [comment]
@orthoxerox 
> implementing the flag as a postfix will affect the most important lexer, the human lexer 😀. If your string is long enough you'll have to scan till its end to find out if the treatment is invariant. It can already be annoying with regexps, but regexps are (usually) short enough for this to work.

Does it really require to 'scan' the string? I'd rather say peek at the end of line ;)

Regex is a reason why I like the idea of postfix. People may be used to this and expect modifiers to be at the end. Also `$i@"..."` looks a bit crowded for me.

**Comment by iam3yal** [comment]
@dstarkowski 

Do you mean `$@i`? because atm the syntax for interpolated strings is `$` and `$@` but I really find it straightforward, also, I'd prefer to have a crowded syntax than to scan or peek at the end what I can see at glance but that's just me. :)

Just because we have modifiers for numbers and regex at the end that in many cases tend to be short doesn't mean we need to follow the same pattern here.

**Comment by jnm2** [comment]
@eyalsk I'd prefer `$i@` because the `i` modifier is modifying the `$` and the `@` is completely orthogonal to them. (Welcome back, long time no see!)

**Comment by rikimaru0345** [comment]
I don't know if adding more prefixes solves anything.
It makes the code very hard to read.

There's also another problem that has to do with localization.

No change: `string.Format("You have {0} things", count);`

With localization support `string.Format(L("You have {0} things"), count);` where L() is a function that does a simple dictionary lookup.

And with `$"You have {count} things"` that is not possible anymore.
There should be a way to get/pass the format-string to a function.

Of course `FormattableString` can solve this as well, but also suffers from the drawback of allocating at least one object.

I have a feeling that both problems can be solved in the same way.
Maybe the real issue here is just that FormattableString is not efficient enough.
Do you agree?

**Comment by jnm2** [comment]
@rikimaru0345 The other really good use case for `$i"Foo {foo}"` is that they would be able to be const string literals.

**Comment by rikimaru0345** [comment]
Maybe the following could solve the performance problems easily and in a fully backwards compatible way. The same code that decides whether to generate FormattableString or a string.Format call could be used to also potentially generate this: 

1. Compile `$"Person: {name}, Age: {years}"` to 

```
struct FormatStringStronglyTyped
{
public readonly string Format;
public readonly string Arg1;
public readonly int Arg2;
}
```

2. Auto implement Deconstruct as well

3. Use tuple deconstruction and projection to call a method that has multiple parameters.

Now you can easily write a method that takes all those parameters from the deconstruction and provide your own formatter like this:

`MyFormattingFunction($"Person: {name}, Age: {years}")
 // Will use deconstruction + projection to deconstruct directly into the parameters of MyFormattingFunction`

`string MyFormattingFunction(string format, string arg1, int arg2)
    => string.Format(format, arg1, arg2); 
// Can set culture settings here, as well as a language localization lookup for 'format'`

Does that idea make sense?

@jnm2 Unfortunately I don't know what can be done about const string literals.
Wouldn't that need to evaluate string.Format() at compile time?

**Comment by jnm2** [comment]
@rikimaru0345 Yes, which is fine if all inputs are constants and the culture is invariant.

**Comment by iam3yal** [comment]
@jnm2 Yup I'm up for whatever syntax tbh as long as they will add it.

p.s. Thanks, been busy with some projects and uni so I'm less active. 😉

**Comment by DmitryMak** [comment]
Allowing shorter symmetrical syntax would be nice from Globalization analyzer perspective: https://github.com/dotnet/runtime/issues/88902

**Comment by stephentoub** [comment]
> C# introduced interpolated string but with no obvious way to decide which CultureInfo to use during formatting.
> Currently it is possible to use InvariantCulture like this:
> FormattableString.Invariant($"count: {count}")
> This requires code to target .NET 4.6 and allocates the FormatabbleString instance instead of just generating call to string.Format.

The `string.Create` method let's you specify a culture, e.g.
```C#
string result = string.Create(CultureInfo.InvariantCulture, $"count: {count}");
```
It utilizes the interpolated string handlers support introduced in C# 10, where the only allocation here will be for the resulting string.

**Comment by DmitryMak** [comment]
@stephentoub I think the current state is not compatible with Globalization rules. Analyzers have to ignore interpolated strings because there is no way to specify the culture explicitly there. Users just have to know that CurrentCulture is used by default and analyzers have to assume it was intentional. I think it is a bit of a gotcha for new users and goes against the goal of the Globalization analyzers (at least as I understand them).

Aside from analyzers, just ergonomically, it feels very strange to have such an asymmetric (for lack of a better word) syntax:

```
string result = $"count: {count}";
string result = string.Create(CultureInfo.InvariantCulture, $"count: {count}");
```

Invariant syntax feels very verbose. Is the assumption that the code that wants to use InvariantCulture should just set CurrentCulture to InvariantCulture (via e.g. Thread.CurrentCulture or System.Globalization.Invariant)?
